File: backend/app/api/voices.py
"""Voice management API routes."""
from typing import List
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from pydantic import BaseModel

from app.api.auth import get_current_user, UserResponse

router = APIRouter()


# Models
class VoiceResponse(BaseModel):
    """Voice clone response."""
    id: str
    name: str
    audio_url: str
    duration_secs: float
    is_default: bool
    status: str
    created_at: datetime


class VoiceCreate(BaseModel):
    """Voice creation request."""
    name: str


# Routes
@router.get("", response_model=List[VoiceResponse])
async def get_voices(current_user: UserResponse = Depends(get_current_user)):
    """Get user's voice clones."""
    # TODO: Get from DB
    return []


@router.post("", response_model=VoiceResponse)
async def create_voice(
    name: str,
    file: UploadFile = File(...),
    current_user: UserResponse = Depends(get_current_user),
):
    """Upload audio to create a voice clone."""
    # TODO: Upload to S3, call Reecho API to clone
    return VoiceResponse(
        id="mock-voice-id",
        name=name,
        audio_url="https://example.com/voice.mp3",
        duration_secs=10.0,
        is_default=False,
        status="ready",
        created_at=datetime.utcnow(),
    )


class VoiceUpdate(BaseModel):
    """Voice update request."""
    name: str | None = None
    is_default: bool | None = None


@router.put("/{voice_id}", response_model=VoiceResponse)
async def update_voice(
    voice_id: str,
    voice_update: VoiceUpdate,
    current_user: UserResponse = Depends(get_current_user),
):
    """Update a voice clone."""
    # TODO: Update in DB
    return VoiceResponse(
        id=voice_id,
        name=voice_update.name or "Updated Voice",
        audio_url="https://example.com/voice.mp3",
        duration_secs=10.0,
        is_default=voice_update.is_default or False,
        status="ready",
        created_at=datetime.utcnow(),
    )


@router.delete("/{voice_id}")
async def delete_voice(
    voice_id: str,
    current_user: UserResponse = Depends(get_current_user),
):
    """Delete a voice clone."""
    # TODO: Delete from DB and S3
    return {"status": "deleted"}


@router.get("/{voice_id}", response_model=VoiceResponse)
async def get_voice(
    voice_id: str,
    current_user: UserResponse = Depends(get_current_user),
):
    """Get a voice clone by ID."""
    # TODO: Get from DB
    return VoiceResponse(
        id=voice_id,
        name="Mock Voice",
        audio_url="https://example.com/voice.mp3",
        duration_secs=10.0,
        is_default=False,
        status="ready",
        created_at=datetime.utcnow(),
    )

File: frontend/src/api/voices.ts
import { getToken } from './auth'

const API_URL = import.meta.env.VITE_API_URL || ''

export interface Voice {
  id: string
  name: string
  audio_url: string | null
  duration_secs: number | null
  is_default: boolean
  status: string
  created_at: string
}

export async function uploadVoice(name: string, file: File): Promise<Voice> {
  const formData = new FormData()
  formData.append('name', name)
  formData.append('file', file)

  const response = await fetch(`${API_URL}/api/voices/upload`, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${getToken()}`,
    },
    body: formData,
  })

  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Upload failed' }))
    throw new Error(error.detail || 'Upload failed')
  }

  return response.json()
}

export async function listVoices(): Promise<Voice[]> {
  const response = await fetch(`${API_URL}/api/voices`, {
    headers: {
      'Authorization': `Bearer ${getToken()}`,
    },
  })

  if (!response.ok) {
    throw new Error('Failed to fetch voices')
  }

  return response.json()
}

export async function deleteVoice(voiceId: string): Promise<void> {
  const response = await fetch(`${API_URL}/api/voices/${voiceId}`, {
    method: 'DELETE',
    headers: {
      'Authorization': `Bearer ${getToken()}`,
    },
  })

  if (!response.ok) {
    throw new Error('Failed to delete voice')
  }
}